### ISSUE TITLE
HttpClient `Default` and `new` implementations may panic too aggressively

### ISSUE THREAD

**Issue by skade** [sagebind/isahc#62]
First of all: I like `isahc` a lot!

https://github.com/sagebind/isahc/blob/42c3a045d9a03b8c6bc6692990468bf85c2bd893/src/client.rs#L314-L338

Documents: 
```
/// Generally such a panic indicates an internal bug or an issue with system
/// configuration. If you need to catch these errors, you can use
/// [`HttpClientBuilder::build`] instead.
```

This confused me.

There are multiple very clear and recoverable conditions here, which should be exposed to the user. One of them is this one:

```
'client failed to initialize: Io(Os { code: 24, kind: Other, message: "Too many open files" })'
```

This is rather easy to hit on a desktop operating system where there's often a file limit of 1024. There's also an obvious mitigation (throttle).

There's 2 decisions here I find a little hard to understand as a user. `Default` is a bit lax described as: 
`Returns the "default value" for a type`. And while that might mean setting up initial state (e.g. for RNG), I'd not assume that most users would expect a call to `default` to panic. Indeed, all implementations in libstd are infallible. I also have a hard time imagining a situation where the isahc client would be used generically through the `Default` interface.

Second, it seems an odd step away from the general Rust strategy of returning `Result` from all operations that may fail as a default way, which also educates and informs users about failures.

I think that `HttpClient::new().unwrap()` would give a better and clearer help to the user, as they see where in their codebase that happens. If `new` returning a `Result` is jarring, it could also be renamed `setup`.

The user story I imagine to be frustrating is the following:
* User develops using `HttpClient::new`. All works in testing.
* A panic of such kind is encountered in deployment
* The software needs to be migrated over to the builder interface to handle
  * This facilitates a feeling of mistrust to the "easy" APIs

Reporting those errors makes the story go as follows:
* User uses `HttpClient::setup()` semantics
* Decides to unwrap the IO Errors
* Figures out that there is a problem that they haven't properly handled
* They change their user code to handle errors better.

**Comment by sagebind** [comment]
Thanks for trying out Isahc, and thanks your thoughtful and well-reasoned issue! You make a fairly convincing argument as to why `Default` should not be implemented. I debated quite a bit whether initialization errors should be glossed over or not, thinking that errors would be quite rare. But max number of file descriptors is a real-world error, as you pointed out, which should not be ignored.

Frankly I copied [reqwest](https://docs.rs/reqwest)'s behavior in this area, but I am now convinced that returning a `Result` is definitely the safer choice. I will include this API change in the next minor release of Isahc.
